Try coveralls

zachmayer opened the pull request:
Added 3 new make targets:
vw_gcov, library_example_gcov, test_gcov which build vw and the examples with GCOV support, then run tests.  This allows coveralls to analyze test coverage in the source code, but slows the tests down signifigantly.  I also edited the travis .yml file to upload the results to coveralls.io and added the badge to the readme.

Someone will need to setup a coveralls account for the main VW project and point the badge in the readme to that badge.  Currently the coveralls badge points only to my fork.

JohnLangford commented:
It looks like you are trying to use autoconf stuff, but for travis we've traditionally used the builtin makefile.  The problem here: 
https://travis-ci.org/JohnLangford/vowpal_wabbit/builds/67037963
does not happen with the builtin makefile, of course.

zachmayer commented:
Wait, which makefile does travis use?  This one? 
https://github.com/JohnLangford/vowpal_wabbit/blob/master/Makefile

JohnLangford commented:
Right.

-John

On Tue, Jun 16, 2015 at 4:21 PM, Zach Mayer <notifications@github.com>
wrote:

> Wait, which makefile does travis use? This one?
> https://github.com/JohnLangford/vowpal_wabbit/blob/master/Makefile
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-112557115>
> .
>

zachmayer commented:
Ok, I tried to add a make target that sets `CXXFLAGS = -g -fprofile-arcs -ftest-coverage` and then makes vw.  Assuming those are the right flags, this might work.   We'll see what travis says.

zachmayer commented:
btw, this is the first makefile I've ever edited in my life, so please excuse my learning curve here... I may be doing very silly things in some of these commits.

I can squash this PR back into a single commit once it's working correctly.

JohnLangford commented:
Travis isn't happy because there is no gcov target.

https://travis-ci.org/JohnLangford/vowpal_wabbit/builds/67579015

zachmayer commented:
Pardon my ignorance, but what's a "target" in a makefile?  I thought this code: 
```{make}
gcov: CXXFLAGS = -g -fprofile-arcs -ftest-coverage
gcov: vw
```
Defines the gcov target (and make gcov works on my local machine).  Perhaps I've formatted something wrong?

zachmayer commented:
Or perhaps I edited the wrong makefile?

JohnLangford commented:
On 06/25/2015 10:01 AM, Zach Mayer wrote:
>
> Pardon my ignorance, but what's a "target" in a makefile? I thought 
> this code:
>
> gcov: CXXFLAGS = -g -fprofile-arcs -ftest-coverage
> gcov: vw
>
> Defines the gcov target (and make gcov works on my local machine). 
> Perhaps I've formatted something wrong?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-115267949>.
>

The basic format for a Makefile is given here:

https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/Makefile#L43

<target>: <dependencies>
     <command>

In this case you have gcov with a dependency on CXXFLAGS, which does not 
work.  Is gcov a file you want to make?  Or is this just supposed to 
trigger building with these flags?

-John

zachmayer commented:
It's supposed to be a trigger for building vw with those flags.

In other words, the gcov target should build the vw target, but with the additional flags.

I'll take another shot at this.

JohnLangford commented:
You need something like:

gcov:
cd vowpalwabbit; CXXFLAGS = -g -fprofile-arcs -ftest-coverage; $(MAKE) 
-j $(NPROCS) things

-John

On 06/26/2015 01:07 PM, Zach Mayer wrote:
>
> It's supposed to be a trigger for building vw with those flags.
>
> In other words, the gcov target should build the vw target, but with 
> the additional flags.
>
> I'll take another shot at this.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-115767669>.
>

zachmayer commented:
gotcha, trying that now

zachmayer commented:
Huh, I got an email saying travis passed:
https://travis-ci.org/zachmayer/vowpal_wabbit/builds/68516136

But the PR says the tests failed

zachmayer commented:
Ohh, I see.  Unit tests are failing on the master branch, so they're also failing for this PR.

JohnLangford commented:
Yup, fixed on the master branch now. (New test was causing problems.)

-John

On 06/26/2015 02:15 PM, Zach Mayer wrote:
>
> Ohh, I see. Unit tests are failing on the master branch, so they're 
> also failing for this PR.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-115822329>.
>

zachmayer commented:
Awesome, thanks.  It seems to be running fine for me now and correctly wired to coveralls, but something else isn't working right.  If you take a look at the [coverage report for one of the builds](https://coveralls.io/builds/2914897) you'll see that coveralls is marking every line of code as "not relevant" and omitting it from the report, which means every codefile has 0 relevant lines of code, which means the coverage isn't running properly.

So now I need to play with the coverage flags a little.  But we're almost there!

zachmayer commented:
Sorry for the flury of commits.  I'm trying to set `CXXOPTIMIZE="-O0 -g"` in the gcov target, e.g. with:
```
cd vowpalwabbit; ./configure --enable-debug; CXXOPTIMIZE="-O0 -g"; CXXFLAGS = "--coverage"; LDFLAGS = "--coverage"; make clean -j 16 things
```

But I get:
```
/bin/sh: 1: ./configure: not found
/bin/sh: 1: CXXFLAGS: not found
/bin/sh: 1: LDFLAGS: not found
```

zachmayer commented:
It doesn't seem like the flags I'm setting are getting passed through make to gcc.

JohnLangford commented:
Configure is incompatible with the default Makefile.

-John

On 06/26/2015 05:19 PM, Zach Mayer wrote:
>
> Sorry for the flury of commits. I'm trying to set |CXXOPTIMIZE="-O0 
> -g"| in the gcov target, e.g. with:
>
> |cd vowpalwabbit; ./configure --enable-debug; CXXOPTIMIZE="-O0 -g"; CXXFLAGS = "--coverage"; LDFLAGS = "--coverage"; make clean -j 16 things
> |
>
> But I get:
>
> |/bin/sh: 1: ./configure: not found
> /bin/sh: 1: CXXFLAGS: not found
> /bin/sh: 1: LDFLAGS: not found
> |
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-115886009>.
>

arielf commented:
In the Makefile (not configure) - I think you should be using this syntax to ensure it crosses process bondaries in all shells:

gcov:
        cd vowpalwabbit && env CXXFLAGS="--flags --you --need.." OTHER_VAR="--other --flags" $(MAKE) things

Leading spaces in command should be TAB

Even better, you can pass them to the $(MAKE) call directly:

      $(MAKE)  CXXFLAGS="--flags ..."  OTHER_VAR="... "   things

You can also add to env vars using $(MAKE) VAR+="more args"

In case of doubt, check the full GNU make documentation:
    http://web.mit.edu/gnu/doc/html/make_toc.html

zachmayer commented:
Ok, I think part of the issue is the defaults flags set `-O3` and I would like to override that to be `-O0`

JohnLangford commented:
Instead of CXXFLAGS, I think you just want to modify FLAGS which is what the child Makefile uses:

https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/Makefile#L32

JohnLangford commented:
(CXXFLAGS is ignored)

zachmayer commented:
Ahah. I'll try directly editing FLAGS.

zachmayer commented:
Ok, setting FLAGS worked for the compilation flags.  I also need to pass flags to the linker (via LDFLAGS I think).  How should I do that?  Setting LDFLAGS didn't seem to work.

zachmayer commented:
All right!  I think it's working end-to-end: https://coveralls.io/builds/2930614
![image](https://cloud.githubusercontent.com/assets/581590/8410027/be4a2330-1e49-11e5-904c-39e937e01bf3.png)

zachmayer commented:
Ok, I think is is all working now!  I squashed down to a single commit. Hopefully travis passes!

zachmayer commented:
Ok now I'm getting what looks like a minor test error:
```{bash}
--- diff -u --minimal train-sets/ref/wiki1K.stderr stderr.tmp
--- train-sets/ref/wiki1K.stderr	2015-06-29 15:32:34.202675000 +0000
+++ stderr.tmp	2015-06-29 15:33:39.486675000 +0000
@@ -5,21 +5,21 @@
 using no cache
 Reading datafile = train-sets/wiki256.dat
 num sources = 1
-average    since         example     example  current  current  current
-loss       last          counter      weight    label  predict features
-10.149311  10.149311           1         1.0  unknown   0.0000      732
-10.369825  10.590340           2         2.0  unknown   0.0000       27
-10.325934  10.282042           4         4.0  unknown   0.0000       53
-10.401774  10.477614           8         8.0  unknown   0.0000       60
-10.356269  10.310764          16        16.0  unknown   0.0000       26
-10.472941  10.589612          32        32.0  unknown   0.0000      125
-10.474854  10.476768          64        64.0  unknown   0.0000      313
-10.425314  10.375773         128       128.0  unknown   0.0000       50
-10.005564  9.585815          256       256.0  unknown   0.0000       33
+average  since         example        example  current  current  current
+loss     last          counter         weight    label  predict features
+10.149309 10.149309            1            1.0  unknown   0.0000      732
+10.369817 10.590324            2            2.0  unknown   0.0000       27
+10.325930 10.282043            4            4.0  unknown   0.0000       53
+10.401775 10.477620            8            8.0  unknown   0.0000       60
+10.356295 10.310814           16           16.0  unknown   0.0000       26
+10.472958 10.589622           32           32.0  unknown   0.0000      125
+10.474864 10.476771           64           64.0  unknown   0.0000      313
+10.425322 10.375779          128          128.0  unknown   0.0000       50
+22.102195 33.779067          256          256.0  unknown   0.0000       33
 
 finished run
 number of examples = 256
-weighted example sum = 256
-weighted label sum = 0
-average loss = 10.0056
+weighted example sum = 256.000000
+weighted label sum = 0.000000
+average loss = 22.102195
 total feature number = 22158
RunTests: test 17: FAILED: ref(train-sets/ref/wiki1K.stderr) != stderr(stderr.tmp)
	cmd: /usr/bin/timeout 30 ../vowpalwabbit/vw -k --lda 100 --lda_alpha 0.01 --lda_rho 0.01 --lda_D 1000 -l 1 -b 13 --minibatch 128 train-sets/wiki256.dat
make: *** [test_gcov] Error 1
```

zachmayer commented:
Unfortunately, it only happens when I make with gcov support, the regular make test still seems to pass

zachmayer commented:
`-ffast-math` was the problem.  Travis is passing but appveyor is failing.  Any suggestions on how to fix the appveyor build?

JohnLangford commented:
If the appveyor _build_ is failing that's a problem, but if it's just the
test cases.... we aren't trusting appveyor at the moment.  Something funky
is going on.

-John


On Mon, Jun 29, 2015 at 2:03 PM, Zach Mayer <notifications@github.com>
wrote:

> -ffast-math was the problem. Travis is passing but appveyor is failing.
> Any suggestions on how to fix the appveyor build?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-116778435>
> .
>

zachmayer commented:
Ok, I think it's just the tests that are failing:
```{bash}
Total tests: 23. Passed: 16. Failed: 4. Skipped: 3.
Test Run Failed.
```

zachmayer commented:
I'll squash back to a single commit, and then I think it's good to go.

Can you create an account on https://coveralls.io and [turn on coveralls for the vowpal_wabbit repositoy](https://coveralls.io/repos/new), so I can point the readme to the correct badge?

zachmayer commented:
Ok, I changed the badge to point at https://coveralls.io/repos/JohnLangford/vowpal_wabbit/badge.svg (Which doesn't exist yet) and squashed back to a single commit.  I'm expecting that travis will pass and apveyor tests will fail.

zachmayer commented:
Cool, travis passes, appveyor doesn't.

zachmayer commented:
This is passing both checks now.  I squashed back to a single commit, so I think it's good to go now.

JohnLangford commented:
Merged in, finally :-)   Let's see how it goes.

zachmayer commented:
Wooo!  Thanks!



—
Sent from Mailbox

On Fri, Jul 31, 2015 at 3:51 PM, John <notifications@github.com> wrote:

> Merged in, finally :-)   Let's see how it goes.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-126796665

zachmayer commented:
@JohnLangford So the last thing to do is to go to https://coveralls.io and create an account. Then click "add repo" and switch the "vowpal wabbit" switch from off to on.

It's very similar to setting up a repo with travis.

![image](https://cloud.githubusercontent.com/assets/581590/9027241/c336eda0-391b-11e5-9690-be18a73b7b0f.png)

zachmayer commented:
@JohnLangford You still need to make a coveralls account and enable access to the VW repo.

zachmayer commented:
@JohnLangford  You still need to make a coveralls account and enable access to the VW repo so the badge on the vowpal wabbit readme will show up correctly.

JohnLangford commented:
Right.  I'm working through overload. Will get to it soon.

-John

On 09/09/2015 09:45 AM, Zach Mayer wrote:
>
> @JohnLangford <https://github.com/JohnLangford> You still need to make 
> a coveralls account and enable access to the VW repo so the badge on 
> the vowpal wabbit readme will show up correctly.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-138913827>.
>

zachmayer commented:
@JohnLangford No worries!  I just didn't want you to forget.  I'll stop bugging you.

JohnLangford commented:
Definitely don't give up on me, and thanks for the reminders.

I've added the coveralls account and enabled access to the VW repo.  What's
next?

-John

On Wed, Sep 9, 2015 at 10:52 AM, Zach Mayer <notifications@github.com>
wrote:

> @JohnLangford <https://github.com/JohnLangford> No worries! I just didn't
> want you to forget. I'll stop bugging you.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-138936340>
> .
>

zachmayer commented:
The next travis build should trigger a coveralls build.  You can either push a dummy commit to master or tell travis to rebuild a recent build.

zachmayer commented:
Looks good now!

JohnLangford commented:
Yep, thanks for your patience.

-John

On Mon, Sep 14, 2015 at 9:30 AM, Zach Mayer <notifications@github.com>
wrote:

> Looks good now!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JohnLangford/vowpal_wabbit/pull/690#issuecomment-140076484>
> .
>
